### ISSUE TITLE
TypeScript doesn't warn when returning an empty object from an overloaded function

### ISSUE THREAD

**Issue by NiGhTTraX** [microsoft/TypeScript#55929]
### 🔎 Search Terms

function overloads object no explicit return type

### 🕗 Version & Regression Information

This is the behavior in every version I tried, and I reviewed the FAQ for entries about function overloads

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABMOcAUA3AhgGxAUwC5EBnKAJxjAHMBKYsEAWwCN9yAoUSWBZVTLgINmbcvVIUq1LuGjwkKdNjxFEjVu0QAfSZRq1EAbw6JE5fFBDkkRgL4cHsngsQss5QauJl9dYkaIWD5SNIgO3PJ87p4qwuqi7BKBLCKa5OHOUUgxXvEaYjp60oYmZhZWNsYOdkA

### 💻 Code

```ts
function foo(value: string): number
function foo(value: number): string
function foo(value: number | string) {
  return {}
}

function bar(value: string): { a: string }
function bar(value: number): { b: number }
function bar(value: number | string) {
  return {}
}
```

### 🙁 Actual behavior

No errors.

### 🙂 Expected behavior

An error that the return type is not assignable to the overloaded return types.

### Additional information about the issue

Returning `42`, `'foo'`, `[]`, `true` or `{ foo: 'bar' }` seems to raise an error correctly.

If I explicitly define the return type of the implementation then TypeScript will correctly raise an error that `{}` is not assignable to any of the overloaded return types. Is it wrong for me to expect that same behavior when I omit the return type?

**Comment by MartinJohns** [comment]
This is working as intended. Since you didn't specify a return type for your implementation signature it's being inferred based on the value you return. Returning `{}` results in the type `{}`, and the type `{}` is both compatible to `number` and `string`, so it's accepted as a valid return type for your implementation.

Duplicate of #39739.

**Comment by DanielRosenwasser** [comment]
Yup, overloads (unfortunately) check for implementation compatibility in a bidirectional manner. The minor addition/correction is that it's `{ a: string }` and `{ b: number }` that are compatible with (assignable to) `{}`.

**Comment by fatcerberus** [comment]
> Yup, overloads (unfortunately) check for implementation compatibility in a bidirectional manner.

To be clear on why this is: The "correct" return type for this implementation is `string | number`, but _that_ only works because both `string` and `number` individually are assignable to `string | number` (just as they are assignable to `{}`)--which incidentally is no less unsound than `{}` because even this lets you return the wrong type for the given argument.

**Comment by fatcerberus** [comment]
> If I explicitly define the return type of the implementation then TypeScript will correctly raise an error that `{}` is not assignable to any of the overloaded return types.

Oh, and this doesn't appear to be the case: [Playground link](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABMOcAUA3AhgGxAUwC5EBnKAJxjAHMBKYsEAWwCN9yAoUSWBZVTLgINmbcvVIUq1LuGjwkKdNjxFEjVu0QAfSZRoSA3gF9EhjokTl8UEOSQmOxjrJ4LELLOUGriZfXTEhohYflI0iM7c8nye3irC6qLsRh4imuSRrjFIcT6JGmI6etJGpuaW1rb2Zs7OQA)

**Comment by MartinJohns** [comment]
> > If I explicitly define the return type of the implementation then TypeScript will correctly raise an error that `{}` is not assignable to any of the overloaded return types.
> 
> Oh, and this doesn't appear to be the case: [Playground link](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABMOcAUA3AhgGxAUwC5EBnKAJxjAHMBKYsEAWwCN9yAoUSWBZVTLgINmbcvVIUq1LuGjwkKdNjxFEjVu0QAfSZRoSA3gF9EhjokTl8UEOSQmOxjrJ4LELLOUGriZfXTEhohYflI0iM7c8nye3irC6qLsRh4imuSRrjFIcT6JGmI6etJGpuaW1rb2Zs7OQA)

I'm fairly sure he meant with the **correct** explicit defined return type, aka `string | number`.

**Comment by NiGhTTraX** [comment]
Thanks to everyone for the replies!

> overloads (unfortunately) check for implementation compatibility in a bidirectional manner

Could you elaborate on this? Do you mean that the implementation is checked contravariantly against the overloads for both arguments and return type, instead of contravariantly for arguments and covariantly for return types? 

> The minor addition/correction is that it's { a: string } and { b: number } that are compatible with (assignable to) {}.

That would explain this...

So I guess the root of the issue comes from TS inferring the return type `{}`, and basically everything is assignable **_to it_**. Moreover, the variance check would also explain the following:

```ts
type Foo = { a: string; b: string; }
type Bar = { b: string; c: string }

function bar(value: string): Foo
function bar(value: number): Bar

function bar(value: number | string) {
  // No error.
  return { b: 'bar' }
}
```

Both `Foo` and `Bar` are assignable to `{ b: string }`, so no error is raised.

While I now understand better why function overloads behave the way they do, I would like to ask if there are any plans to improve them, perhaps as part of the `strictFunctionTypes` flag, or a new similar flag?

**Comment by MartinJohns** [comment]
> Both `Foo` and `Bar` are assignable to `{ b: string }`, so no error is raised.

That's simply because the implementation is not checked if it's with the overloads compatible. See #13235.

**Comment by NiGhTTraX** [comment]
> For the issue with the inferred {} there's surely an eslint rule preventing this.

There is a [rule](https://typescript-eslint.io/rules/ban-types/) against using the `{}` type explicitly, but it doesn't work for inferred types.

> > Both `Foo` and `Bar` are assignable to `{ b: string }`, so no error is raised.
> 
> That's simply because the implementation is not checked if it's with the overloads compatible. See #13235.

I'm not sure I understand, and I can't see how the linked issue is related. And the more I think about it, the less I understand why TS doesn't raise an error when inferring `{ b: string }`, because it does raise one when declared explicitly ([playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYg9nKBeKBvKBDAXFAzsAJwEsA7AcwG4oAjHfY8qgXwChRIoAhDA5NGuoVKUoAE0EMyUViwBmAVxIBjYETgkaPABQA3DABt5ECcICUOeHDmKVajdW17DxqCXkBbahALmuPa8qq6poEugZGOG6e3lAAPnhC5KZoLFBQAPTpUAByiN4EcAQAdKlQBBDA8gQa6LRQAOQOBPXSLDKZUAAqABZEuFBwOt76cBiieERkJBiV5VB9rnDAUEpw7mAzRNT60ADuRMDd88D9ROs77hAkwJvBuJPTsxBFWgBMAMwAnAAspgG2wQcAC8ws4TEkLAh-kF7BgQU4Iq4PF4fDhuARoXZNPDwi4oii4glJL5ajhGjwWkwUmlyrMagIGk1KW0gA)):

```ts
type Foo = { a: string; b: string; }
type Bar = { b: string; d: string }

function bar(value: string): Foo
function bar(value: number): Bar
function bar(value: number | string) {
  // No error.
  return { b: 'bar' }
}

// This overload signature is not compatible with its implementation signature.(2394)
function baz(value: string): Foo
function baz(value: number): Bar
function baz(value: number | string): { b: 'bar' } {
  return { b: 'bar' }
}
```

So there is some type checking to make sure the type of the implementation satisfies the overloads, but if the type is inferred things get wonky?

**Comment by fatcerberus** [comment]
@NiGhTTraX The “no error” case is actually inferred as `{ b: string }` due to widening, which is a strict supertype of both `Foo` and `Bar`.  That’s *not* the case for `{ b: "Bar" }`.  If you add an `as const` to that return, I bet you get the same error as `baz`.

**Comment by NiGhTTraX** [comment]
> the implementation is accepted if its return type is a strict supertype of all the overloaded return types

Gotcha, thanks!
